SSA/Projects/CIRCT/comb/comb.lean

```diff
@@ -0,0 +1,560 @@
+  | .shru w => [Ty.bv w, Ty.bv w]
+  | .sub w => [Ty.bv w, Ty.bv w]
+  | .xor w => [Ty.bv w, Ty.bv w]
+
+-- return type CONF
+@[simp, reducible]
+def Op.outTy : Op  → Ty
+  | .add w => Ty.bv w
+  | .and w => Ty.bv w
+  | .concat w₁ w₂ => Ty.bv (w₁ + w₂)
+  | .divs w => Ty.bv w
+  | .divu w => Ty.bv w
+  | .extract w n => Ty.bv (w - n)
+  | .icmp _ _ => Ty.bool
+  | .mods w => Ty.bv w
+  | .modu w => Ty.bv w
+  | .mul w => Ty.bv w
+  | .mux w => Ty.bv w
+  | .or w =>  Ty.bv w
+  | .parity _ => Ty.bool
+  | .replicate w n =>  Ty.bv (w * n)
+  | .shl w =>  Ty.bv w
+  | .shrs w =>  Ty.bv w
+  | .shru w =>  Ty.bv w
+  | .sub w =>  Ty.bv w
+  | .xor w => Ty.bv w
+
+@[simp, reducible]
+def Op.signature : Op → Signature (Ty) :=
+  fun o => {sig := Op.sig o, outTy := Op.outTy o, regSig := []}
+
+instance : DialectSignature Comb := ⟨Op.signature⟩
+
+@[simp]
+instance : DialectDenote (Comb) where
+    denote
+    | .add _, arg, _ => Comb.add (arg.getN 0 (by simp [DialectSignature.sig, signature])) (arg.getN 1 (by simp [DialectSignature.sig, signature]))
+    | .and _, arg, _ => Comb.and (arg.getN 0 (by simp [DialectSignature.sig, signature])) (arg.getN 1 (by simp [DialectSignature.sig, signature]))
+    | .concat _ _, arg, _ => Comb.concat (arg.getN 0 (by simp [DialectSignature.sig, signature])) (arg.getN 1 (by simp [DialectSignature.sig, signature]))
+    | .divs _, arg, _ => Comb.divs (arg.getN 0 (by simp [DialectSignature.sig, signature])) (arg.getN 1 (by simp [DialectSignature.sig, signature]))
+    | .divu _, arg, _ => Comb.divu (arg.getN 0 (by simp [DialectSignature.sig, signature])) (arg.getN 1 (by simp [DialectSignature.sig, signature]))
+    | .extract w n, arg, _ => Comb.extract (arg.getN 0 (by simp [DialectSignature.sig, signature])) n
+    | .icmp _ _, arg, _ => Comb.icmp (arg.getN 0 (by simp [DialectSignature.sig, signature])) (arg.getN 1 (by simp [DialectSignature.sig, signature])) (arg.getN 2 (by simp [DialectSignature.sig, signature]))
+    | .mods _, arg, _ => Comb.mods (arg.getN 0 (by simp [DialectSignature.sig, signature])) (arg.getN 1 (by simp [DialectSignature.sig, signature]))
+    | .modu _, arg, _ => Comb.modu (arg.getN 0 (by simp [DialectSignature.sig, signature])) (arg.getN 1 (by simp [DialectSignature.sig, signature]))
+    | .mul _, arg, _ => Comb.mul (arg.getN 0 (by simp [DialectSignature.sig, signature])) (arg.getN 1 (by simp [DialectSignature.sig, signature]))
+    | .mux _, arg, _ => Comb.mux (arg.getN 0 (by simp [DialectSignature.sig, signature])) (arg.getN 1 (by simp [DialectSignature.sig, signature])) (arg.getN 2 (by simp [DialectSignature.sig, signature]))
+    | .or _, arg, _ => Comb.or (arg.getN 0 (by simp [DialectSignature.sig, signature])) (arg.getN 1 (by simp [DialectSignature.sig, signature]))
+    | .parity _, arg, _ => Comb.parity (arg.getN 0 (by simp [DialectSignature.sig, signature]))
+    | .replicate _ n, arg, _ => Comb.replicate (arg.getN 0 (by simp [DialectSignature.sig, signature])) n
```

Comment:
This is effectively how you model `replicate` already; you've ignored the dynamic argument in favour of the `n` annotation